File: src/pipeline/registry.ts
/**
 * Stage registry. Stages register themselves by name so a chain can be
 * described as data – a list of names and params from a CLI flag, a config
 * file or the UI – rather than as imports wired up in code.
 */

import type { Stage } from "./types";

const stages = new Map<string, Stage<object, unknown>>();

export function registerStage(stage: Stage<object, unknown>): void {
  if (stages.has(stage.name)) {
    throw new Error(`Stage "${stage.name}" is already registered`);
  }
  stages.set(stage.name, stage);
}

export function getStage(name: string): Stage<object, unknown> {
  const stage = stages.get(name);
  if (!stage) {
    const known = [...stages.keys()].join(", ") || "none";
    throw new Error(`Unknown stage "${name}" (registered: ${known})`);
  }
  return stage;
}

export function hasStage(name: string): boolean {
  return stages.has(name);
}

export function registeredStages(): Stage<object, unknown>[] {
  return [...stages.values()];
}

File: crates/leveller-pipeline/src/registry.rs
//! The stage registry.
//!
//! Stages are held by name so a chain can be described as data — a list of
//! names and parameters from a command-line flag, a config file or the UI —
//! rather than as imports wired up in code.
//!
//! A registry is built rather than global, which is what lets the trained
//! denoise backend be present in one build and absent in another without every
//! caller learning about it.

use std::sync::Arc;

use crate::stage::Stage;

#[derive(Clone, Default)]
pub struct Registry {
    stages: Vec<Arc<dyn Stage>>,
}

impl Registry {
    pub fn new() -> Self {
        Self::default()
    }

    /// Add a stage. Registration order is the chain's default order.
    ///
    /// # Panics
    /// If a stage of that name is already registered — two stages answering to
    /// one name is a wiring mistake, not a runtime condition.
    pub fn register(&mut self, stage: Arc<dyn Stage>) -> &mut Self {
        assert!(
            self.get(stage.name()).is_none(),
            "stage \"{}\" is already registered",
            stage.name()
        );
        self.stages.push(stage);
        self
    }

    pub fn get(&self, name: &str) -> Option<&Arc<dyn Stage>> {
        self.stages.iter().find(|s| s.name() == name)
    }

    pub fn contains(&self, name: &str) -> bool {
        self.get(name).is_some()
    }

    /// Every stage, in registration order.
    pub fn stages(&self) -> &[Arc<dyn Stage>] {
        &self.stages
    }

    /// Every stage's name, in registration order — which is chain order.
    pub fn names(&self) -> Vec<&'static str> {
        self.stages.iter().map(|s| s.name()).collect()
    }

    pub fn is_empty(&self) -> bool {
        self.stages.is_empty()
    }

    pub fn len(&self) -> usize {
        self.stages.len()
    }
}
